README.md
# python-openimu

A message communication tool for OpenIMU, OpenRTK and other devices of Aceinna

## Working Environment 
- Windows10: python 3.7
- Mac OS: python 3.7

## Steps

### 1. Start the tool
There are 3 ways to run the tool

#### Prepare
1. Install the dependency library. It is better to create a virtual environment before to do it.

```bash
$ pip install -r requirements.txt
```

2. Install pcap api library(`optional`). Because we try to support INS401 a new device from Aceinna. It would use 100BASE-T1 ethernet, and the data transfer is in low-level of network, so we need help of pcap api. Actually, we integrated `scapy` in our project. The pcap api library is the dependency of `scapy`. If you are using a INS401 device, please also install the library based on your platform.

| Platform | Libraray | Reference |
| - | - | - |
| Windows | `npcap` | https://scapy.readthedocs.io/en/latest/installation.html#windows |
| Ubuntu | `tcpdump` | https://scapy.readthedocs.io/en/latest/installation.html#debian-ubuntu-fedora |
| Mac | `libpcap` | https://scapy.readthedocs.io/en/latest/installation.html#mac-os-x |


#### A. From source code

##### Run
Please use this way if you want to contribute the project.
```bash
$ python main.py
```
#### B. Work as execution file

##### Build
The executable will be generated in `dist` folder.
```bash
$ pyinstaller build.spec
```

##### Run
```bash
$ cd dist
$ ./ans-devices
```

#### C. From pip

##### Install
```bash
$ pip install openimu
```

##### Run
```
$ openimu
```

#### Startup Arguments
You can specify some arguments while run the tool

Arguments:

| Name | Type | Default | Description |
| - | :-: | :-: | - |
| -i, --interface | String | 'default' | Value should be `uart`, `eth`, `100base-t1` |
| -p, --port | Number | '8000' | Value should be an available port |
| --device-type | String | 'auto' | Value should be one of `IMU`, `RTK`, `DMU` |
| -b, --baudrate | String | None | Value should be a valid baudrate. The valid value should be one of `38400`, `57600`, `115200`, `230400`, `460800` |
| -c, --com-port | String | 'auto' | Value should be a COM port |
| --console-log | Boolean | False | Output log on console |
| --debug | Boolean | False | Log debug information |
| --with-data-log | Boolean | False | Contains internal data log (OpenIMU only) |
| -s, --set-user-para | Boolean | False | Set uesr parameters (OpenRTK only) |


### 2. Connect Aceinna device
Link device to your pc or mac. The tool will auto detect the linked device by default.

[More Usage](USAGE.md "More Usage")

## Work Mode
### Default Mode
Normally, python-openimu works as default mode. It will establish a websocket server, then exchange messages through the websocket protocol. And it should work with [aceinna developers site](https://developers.aceinna.com "Aceinna Developers Site"), it allows user to do data monitor, configuration and firmware management.

### Command Line Mode
You can specify the startup argument `--cli` to switch to Command Line Mode. Command Line Mode helps you interact with device directly. And it also supply command to start a websocket server, so that you can use the full features of Default Mode. 

Command Line Mode supports some commands for using, below is a list of commands description,

#### Help
Show help menu. It would show a list of description for all supported commands.
```bash
$ help
```

#### Get Device Info
Show information of connected device.
```bash
$ connect
```

#### Get Parameter (OpenIMU Only)
Retrieve current value of specified parameter.
```bash
$ get param_name
```

#### Set Parameter (OpenIMU Only)
Update the value of specified parameter. The value would be recoverd after device power off.
```bash
$ set param_name param_value
```

#### Save Configuration
Save the configuration into EEPROM. The value would be permanently saved.
```bash
$ save
```

#### Record Data (OpenIMU Only)
Log the device output data in path /data. It is not supported for OpenRTK, because OpenRTK device will automatically log data when it is connected. 
```bash
$ record
```

#### Upgrade Firmware
Upgrade firmware from a specified path. The binary file should match with the device. This is a high risk command.
```bash
$ upgrade path/to/bin
```

#### Start Server
Establish a websocket server.
```bash
$ server_start
```

#### Stop Server
Stop the websocket server. If there is websocket server runing, it has to stop it when you want to use other command.
```bash
$ stop
```

#### Exit
Quit from Command Line Mode
```bash
$ exit
```


## Interface
Aceinna Device could be connected with your PC via UART or LAN. The supported interface is depended on the device type.
| Device Type | Supported Protocols | Description |
| - | - | - |
| DMU | `uart` | |
| OpenIMU | `uart` | |
| OpenRTK | `uart`, `eth` | The startup argument `-i eth` is supported |
| RTK330L | `uart` |  |
| INS401 | `100base-t1` | The startup argument `-i 100base-t1` is supported |


## Parse Tool
There is a log parse tool integrated in. It could parse `OpenRTK`, `RTK330LA`, `INS401` raw data log from data folder. Assgin `parse` to start it.

### Arguments:

| Name | Type | Default | Description |
| - | :-: | :-: | - |
| -t | String | 'openrtk' | Switch work mode. Value should be one of `openrtk`,`rtkl`,`ins401` |
| -p | String | '.' | Value should be a valid path. It could be the container folder of log files |
| -i | Number | 5 | INS kml rate(hz) |

### Example

Run from source code
```bash
$ python main.py parse
```

Work as execution file
```bash
$ ans-devices parse
```

Run from pip
```bash
$ openimu parse
```

## Changelogs and Release Notes

Please refer to [HISTORY.md](HISTORY.md "Change History")


HISTORY.md
# History

---
## 2.7.0 / 2023-3-31
- [OpenRTK] Add CAN port parameter config in command line mode.
- [OpenRTK] Add SN check logic for different version of OpenRTK (BA or BX).
- [OpenRTK] Optimize the connection under lan port.
- [RTK330LA] Update bootloader array of 8100 upgrade worker and 9100 upgrade worker.

## 2.6.1 / 2022-02-23
- [OpenRTK] Support ST8100Bx chip upgrade.
- [OpenRTK/RTK330LA] Resolve cannot ping device while upgrading.
- [DMU] Resolve ping issue in baudrate 230400.
- [Framework] Update native library dependency.

## 2.6.0 / 2021-10-11
- [INS401] Support communication with ethernet 100base-t1 protocol.

## 2.5.0 / 2021-07-02
- [DMU] Support firmware upgrade.
- [OpenRTK/RTK330LA] Support ST9100 chip upgrade.
- [Framework] Fix NTRIP client disconnect issue.
- [Framework] Refactor upgrade worker.
- [Framework] Integrate rtk log parser.
- [Framework] Update startup parameters.

## 2.4.0 / 2021-05-07
- [DMU] Resolve firmware version from ID message.
- [DMU] Support record data in developers site.
- [DMU/IMU] Support GF, SF, RF, WF for uart command run tool in developers site.
- [OpenRTK/RTK330LA] Log device info when device is connected.

## 2.3.2 / 2021-04-12
- [DMU] Support INS330BI
- [RTK330LA] Fix the wrong GNSS port logging

## 2.3.1 / 2021-04-02
- [OpenIMU] Update openimu.json for OpenIMU300RI
- [OpenRTK] Update GGA parser
- [OpenRTK] Split log name of RTK330LA from OpenRTK
- [Framework] Optimize the device information match
- [Framework] Support force to bootloader

## 2.3.0 / 2021-02-28
- [OpenIMU] Support packet data statistics.
- [OpenRTK] Support RTK330L.
- [OpenRTK] Save predefined parameters when connected.
- [Framework] Adjust the json file location. Different product would have single application json file.
- [Framework] Support upgrade firmware through aceinna developers site.
- [Framework] Enhance the upgrade center.
- [Framework] Refactor WebServer, consider websocket server is a message tunnel.
- [Framework] Refactor bootstrap, to make it clear for later maintainer.

## 2.2.4 / 2020-12-18
- [OpenRTK] Remove console print and add print.log to save these infomation.
- [OpenRTK] Update openrtk parse to make kml files

## 2.2.3 / 2020-12-01
- [OpenRTK] Update Configuration read size
- [Framework] Fix cannot parse 'sC', 'uB' command

## 2.2.2 / 2020-11-26
- [OpenIMU] Add exception handler when log data, although file is closed
- [OpenIMU] Add uC,uA,gA command response
- [OpenRTK] Fix upgrade issue through web

## 2.2.1 / 2020-11-17

- [OpenIMU] Fix the mag align command cannot correctly response.
- [Framework] Update the usage of asyncio.
- [Framework] Fix cannot connect the websocket server on some versions of windows.
- [Framework] Support to start executor as a cli tool with startup parameter `--cli`.
- [Framework] Fix data log will auto start after firmware upgrade without setting auto start.

## 2.2.0 / 2020-11-9

- [OpenRTK] Important update for INS App v23.00, can't suitable for v2.0.0 or v20.00.
- [OpenRTK] Modify user data packets.
- [OpenRTK] Log base rtcm data on debug port.
- [DMU] Add A1,A2 packet response for DUM device.
- [Framework] Add Unit test cases.

## 2.1.7 / 2020-08-26

- [OpenRTK] Print 'NMEA' and #INSPVA.
- [Framework] Improved the ping perform on devices.

## 2.1.6 / 2020-08-19

- [OpenRTK] Modify INS json: suitable for INS_APP v2.0.0.
- [Framework] Improve the output of console.

## 2.1.5 / 2020-08-11

- [Framework] Support download combined GNSS_RTK_SDK and INS_APP Firmware.
- [Framework] Display python driver version.
- [Framework] Remove upgrade file when upgrade firmware.

## 2.1.4 / 2020-07-23

- [OpenRTK] Add 'rD' command to restore OpenRTK330 default configuration.
	User can find 'RESTORE DEFAULTS' button in OpenRTK->SETTINGS.
- [OpenRTK] Add 'gB' command to get configuration according to range of parameterID.
- [OpenRTK] Support update GNSS_RTK_SDK in App Center.
- [Framework] Enhance the message parser from device.



commands.py
"""
Python command line interface
Date: 07-05-2018
Author:
"""

import serial
import string
import time
import json
import time
import threading
import math
import os
import sys
import tornado.websocket
import tornado.ioloop
import tornado.httpserver
import tornado.web
import psutil

from imu_input_packet import InputPacket
from server import WSHandler
from global_vars import imu
from os import getpid

class OpenIMU_CLI:
    def __init__(self):
        '''Initialize command line interface
        '''
        self.command = "help"         # set default command to help
        self.input_string = []        # define input_string
        self.cli_properties = [] 
        self.current_command =[]
        self.baud_rate = 0
        self.accel_lpf = 0
        self.rate_lpf = 0
        self.orien = 0
        self.thread = threading.Thread(target=self.start_tornado)
        self.http_server = None
        self.http_server_running = False

    def help_handler(self):
        print("Usage: ")
        for x in self.cli_properties:
            print(x['name'] + " : " + x['description']) 
        
        return True

    def connect_handler(self):
        '''connect command is used to find OpenIMU device
        '''
        imu.find_device()

        return True

    def upgrade_handler(self):
        '''upgrade command is used for firmware upgrade and followed by file name
        '''
        input_args = len(self.input_string)
        if (input_args == 1):
            print("Usage:")
            print("upgrade file_name")
        else:
            file_name = self.input_string[1]
            imu.ws = False
            if file_name is not " ":
                if imu.openimu_upgrade_fw_prepare(file_name) == True:
                    while not imu.openimu_finish_upgrade_fw():
                        imu.openimu_upgrade_fw(file_name)
                    imu.openimu_start_app()
                else:
                    print("Error: file doesn't exist")  
        return True

    def record_handler(self):
        '''record command is used to save the outputs into local machine
        '''
        imu.ws = False
        imu.start_log()
        return True

    def stop_handler(self):
        '''record command is used to save the outputs into local machine
        '''
        if imu.logging == 1 or self.http_server_running == True:
            if imu.logging == 1:
                imu.stop_log()
            if self.http_server_running ==True:
                self.server_stop()
        else:
            print("no ongoing recording or server not running now!")
        time.sleep(2)
        return True

    def run_handler(self):
        '''used by customers
        '''
        print("Running...")
        self.server_start_handler()
        return True

    def get_handler(self):
        '''get configuration or data from OpenIMU device
        '''
        input_args = len(self.input_string)
        param_properties = imu.imu_properties['userConfiguration']
 
        if (input_args == 1):
            print("Usage: get [options]")
            print("Option: ")
            i = 2
            while i < len(param_properties):
                print(param_properties[i]['argument'])
                i += 1
            return True
        else:
            i = 2
            while i < len(param_properties):
                x = param_properties[i]
                if (x['argument'] == self.input_string[1]):
                    break
                i += 1
                if (i == len(param_properties)):
                    print("Usage: get [options]")
                    print("Option: ")
                    i = 2
                    while i < len(param_properties):
                        print(param_properties[i]['argument'])
                        i += 1
                    return True
        
        param = imu.openimu_get_param(x['paramId'])
        return True
   
    def set_param(self, command, value):
        param_properties = imu.imu_properties['userConfiguration']
        i = 2
        while i < len(param_properties):
            x = param_properties[i] 
            if (x['argument'] == command):
               break
            i += 1

        if i < len(param_properties):
            if command == "type" or command == "orien":
                imu.openimu_update_param(x['paramId'], value)
            else:
                imu.openimu_update_param(x['paramId'], int(value))
        else:
            print("Usage: set [options] <value>") 
            
        return True
     
    def set_handler(self):
        '''set parameters' values
        '''
        input_args = len(self.input_string)
        param_properties = imu.imu_properties['userConfiguration']

        if input_args == 1:
            print("Usage: set <options> <values>")
            i = 2
            while i < len(param_properties):
                x = param_properties[i] 
                print(x['argument'])
                i += 1
            return True
        else:
            i = 2
            while i < len(param_properties):
                x = param_properties[i] 
                if (x['argument'] == self.input_string[1]):
                    break
                i += 1

        if input_args == 2:
            if i == len(param_properties):
                print("Usage: set <options> <values>")
                i = 2
                while i < len(param_properties):
                    x = param_properties[i] 
                    print(x['argument'])
                    i += 1
            else:
                print("Usage: set " + x['argument'] + " <values>")
                print("values: ") 
                print(x['options'])
            return True

        if ((x['type'] == "char8" and self.input_string[2] not in x['options']) or 
            (x['type'] == "int64" and int(self.input_string[2]) not in x['options'])):
                print("Usage: set " + x['argument'] + " <values>")
                print("values: ") 
                print(x['options'])
                return True
        else:
            self.set_param(self.input_string[1], self.input_string[2])

        return True

    def save_handler(self):
        C = InputPacket(imu.imu_properties, 'sC')
        imu.write(C.bytes)
        return True
   
    def start_tornado(self):
        application = tornado.web.Application([(r'/', WSHandler)])
        self.http_server = tornado.httpserver.HTTPServer(application)
        if sys.version_info[0] > 2:
            import asyncio
            asyncio.set_event_loop(asyncio.new_event_loop())
        self.http_server.listen(8000)
        tornado.ioloop.IOLoop.instance().start()
         
    def server_start_handler(self):
        imu.ws = True
        self.thread.start()
        self.http_server_running = True
        return True

    def server_stop(self):
        print("server stops")
        if self.http_server is not None:
           self.http_server.stop()
        imu.ws = False
        self.http_server_running = False
        pid = getpid()
        p = psutil.Process(pid)
        p.kill() 

    def command_handler(self):
        '''main routine to handle input command
        '''       
        self.cli_properties = imu.imu_properties['CLICommands']

        while True:
            if sys.version_info[0] < 3:
                token = raw_input(">>")
            else:
                token = input(">>")
            self.input_string = token.split(" ")
             
            if token.strip() == 'exit':
                if self.http_server_running == True:
                    self.http_server_running = False
                    self.server_stop()
                else:
                    break
            if self.http_server_running == True and token.strip() != 'stop':
                print("server is on-going, please stop it")
                continue

            for x in self.cli_properties:
                if x['name'] == self.input_string[0]:
                    self.current_command = x
                    eval('self.%s()'%(x['function']))
                    break
            else:
                self.help_handler()    

        return True 

if __name__ == "__main__":
    imu.find_device()
    cli = OpenIMU_CLI()
    cli.command_handler()
